go/libraries/doltcore/sqle/dtables/unscoped_diff_table.go

```diff
@@ -0,0 +1,227 @@
+}
+
+// NewUnscopedDiffTableItr creates a UnscopedDiffTableItr from the current environment.
+func NewUnscopedDiffTableItr(ctx *sql.Context, ddb *doltdb.DoltDB, head *doltdb.Commit) (*UnscopedDiffTableItr, error) {
+	commits, err := actions.TimeSortedCommits(ctx, ddb, head, -1)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &UnscopedDiffTableItr{ctx, ddb, commits, 0, nil, -1}, nil
+}
+
+// HasNext returns true if this UnscopedDiffItr has more elements left.
+func (itr *UnscopedDiffTableItr) HasNext() bool {
+	// There are more diff records to iterate over if:
+	//   1) there is more than one commit left to process, or
+	//   2) the tableNames array isn't nilled out and has data to process
+
+	return itr.commitIdx+1 < len(itr.commits) || itr.tableNames != nil
+}
+
+// Next retrieves the next row. It will return io.EOF if it's the last row.
+// After retrieving the last row, Close will be automatically closed.
+func (itr *UnscopedDiffTableItr) Next(*sql.Context) (sql.Row, error) {
+	if !itr.HasNext() {
+		return nil, io.EOF
+	}
+
+	defer func() {
+		// Increment the table name index, and if it's the end of the table names array,
+		// move to the next commit and reset the table name index
+		itr.tableNameIdx++
+		if itr.tableNameIdx >= len(itr.tableNames) {
+			itr.tableNameIdx = -1
+			itr.tableNames = nil
+			itr.commitIdx++
+		}
+	}()
+
+	// Load table names if we don't have them for this commit yet
+	for itr.tableNames == nil {
+		err := itr.loadTableNames(itr.commits[itr.commitIdx])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	commit := itr.commits[itr.commitIdx]
+	hash, err := commit.HashOf()
+	if err != nil {
+		return nil, err
+	}
+
+	meta, err := commit.GetCommitMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	return sql.NewRow(hash.String(), meta.Name, meta.Email, meta.Time(),
+		meta.Description, itr.tableNames[itr.tableNameIdx]), nil
+}
+
+// loadTableNames loads the set of changed tables for the current commit into this iterator, taking
+// care of advancing the iterator if that commit didn't mutate any tables and checking for EOF condition.
+func (itr *UnscopedDiffTableItr) loadTableNames(commit *doltdb.Commit) error {
+	tableNames, err := itr.calculateChangedTables(commit)
+	if err != nil {
+		return err
+	}
+
+	// If there are no table deltas for this commit (e.g. a "dolt doc" commit),
+	// advance to the next commit, checking for EOF condition.
+	if len(tableNames) == 0 {
+		itr.commitIdx++
+		if !itr.HasNext() {
+			return io.EOF
+		}
+	} else {
+		itr.tableNames = tableNames
+		itr.tableNameIdx = 0
+	}
+
+	return nil
+}
+
+// calculateChangedTables calculates the tables that changed in the specified commit, by comparing that
+// commit with its immediate ancestor commit.
+func (itr *UnscopedDiffTableItr) calculateChangedTables(commit *doltdb.Commit) ([]string, error) {
+	toRootValue, err := commit.GetRootValue()
+	if err != nil {
+		return nil, err
+	}
+
+	parent, err := itr.ddb.ResolveParent(itr.ctx, commit, 0)
```

Comment:
Nit: you're doing more work than necessary here, since you're already going to resolve the parent when you iterate over it. Fine for a first pass though.